Source/DafnyRuntime/DafnyRuntimeRust/src/lib.rs

```diff
@@ -30,275 +30,1335 @@ pub type SizeT = usize;
-            boxed.replace(Rc::clone(
-                &Sequence::<T>::new_array_sequence(&result)));
-            result
-        } else {
-            panic!("This should never happen")
+where T: DafnyType {
+    pub fn from_array(values: &Vec<T>) -> Sequence<T> {
+        Sequence::ArraySequence {
+            values: Rc::new(values.clone())
+        }
+    }
+    pub fn from_array_slice(values: &Vec<T>, start: &DafnyInt, end: &DafnyInt) -> Sequence<T> {
+        Sequence::ArraySequence {
+            values: Rc::new(values[start.to_usize().unwrap()..end.to_usize().unwrap()].to_vec())
+        }
+    }
+    pub fn from_array_take(values: &Vec<T>, n: &DafnyInt) -> Sequence<T> {
+        Sequence::ArraySequence {
+            values: Rc::new(values[..n.to_usize().unwrap()].to_vec())
+        }
+    }
+    pub fn from_array_drop(values: &Vec<T>, n: &DafnyInt) -> Sequence<T> {
+        Sequence::ArraySequence {
+            values: Rc::new(values[n.to_usize().unwrap()..].to_vec())
+        }
+    }
+    pub fn from_array_owned(values: Vec<T>) -> Sequence<T> {
+        Sequence::ArraySequence {
+            values: Rc::new(values),
+        }
+    }
+    pub fn new_concat_sequence(left: &Sequence<T>, right: &Sequence<T>) -> Sequence<T> {
+        Sequence::ConcatSequence {
+            left: Rc::new(UnsafeCell::new(left.clone())),
+            right: Rc::new(UnsafeCell::new(right.clone())),
+            length: left.cardinality_usize() + right.cardinality_usize(),
+            boxed: Rc::new(RefCell::new(None)),
+        }
+    }
+    pub fn to_array(&self) -> Rc<Vec<T>> {
+        // Let's convert the if then else below to a proper match statement
+        match self {
+            Sequence::ArraySequence { values, .. } =>
+              // The length of the elements
+              Rc::clone(values),
+            Sequence::ConcatSequence { length, boxed, left, right } => {
+                let clone_boxed = boxed.as_ref().clone();
+                let clone_boxed_borrowed = clone_boxed.borrow();
+                let borrowed: Option<&Rc<Vec<T>>> = clone_boxed_borrowed.as_ref();
+                if let Some(cache) = borrowed.as_ref() {
+                    return Rc::clone(cache);
+                }
+                // Let's create an array of size length and fill it up recursively
+                // We don't materialize nested arrays because most of the time they are forgotten
+                let mut array: Vec<T> = Vec::with_capacity(*length);
+                Sequence::<T>::append_recursive(&mut array, self);
+                let result = Rc::new(array);
+                let mut cache = boxed.borrow_mut();
+                let mutable_left: *mut Sequence<T> = left.get();
+                let mutable_right: *mut Sequence<T> = right.get();
+                unsafe { *mutable_left = seq!() };
+                unsafe { *mutable_right = seq!() };
+                *cache = Some(result.clone());
+                result
```

Comment:
Is this approach as thread-safe as the older version with LazySequence and atomic boxing?

Reply:
The answer is yes, because it was not thread-safe previously anyway.
Rc is a reference counting that is not thread-safe. Generating concurrent Rust code will require to change every Rc to Arc, which has locking but a lot of overhead for locking.